### DISCUSSION TITLE
How to include graphics when using `tar_render()`? Error: `Cannot find the file(s)`

### DISCUSSION THREAD

**Discussion by RHReynolds** [ropensci/tarchetypes#134] [Help]
Hi,

I would like `tar_render()` to render an `.rmd` that includes several external graphics that are stored in a directory in the same location where the `.rmd` is. However, I keep getting the error `Cannot find the file(s): "figures/example.png"`.

The way I am currently running this is:

- In an RStudio Project, which is where the targets pipeline is defined
- Relative to the project directory, the `.rmd` is stored in the directory, `./docs/`, while figures are stored in `./docs/figures`
- I have defined the `./docs/figures` in my `_targets.R` using:
```
tar_target(
        name = fig_dir,
        command = here::here("docs", "figures"),
        format = "file"
        )
```
- This target is passed to `tar_render()`, using the following command:
```
tar_render(
        name = report,
        path = here::here("docs", "example.rmd"),
        params =
            list(
                ext_figures = fig_dir
            )
    )
```
- Within the `.rmd`, I use tar_read in the yaml header to call the path to the target, `fig_dir` (see below).
```
params:
    ext_figures: !r targets::tar_read(fig_dir, store = here::here("_targets"))
```
- Finally, the figure is called within the rmd using: `knitr::include_graphics(file.path(params$ext_figures, "example.png"))
`

Appreciate this is quite a long-winded solution, but it was important that I was (i) able to track the dependencies going into the `.rmd`, as in more complicated cases, I will be using the outputs of other processes as input to my `.rmd` and (ii) still able to knit the `.rmd` using RStudio's knit button. The above solution allows me to knit the `.rmd` without issues. But unfortunately it doesn't seem to work with `tar_render()`. Do you have any suggestions as to how to resolve this?

**Comment by wlandau** [comment]
Seems similar to https://github.com/ropensci/tarchetypes/discussions/116#discussioncomment-4151828. Experimenting with regular Markdown syntax and `withr::with_dir(knitr::include_graphics(...))` may help. Other than that, it would be easier to help with a [runnable reprex](https://books.ropensci.org/targets/help.html#write-a-reprex).

**Comment by RHReynolds** [comment]
Thanks for the quick response and suggestions! 

I managed to get the following to work both with `tar_make()` and RStudio knit:  `withr::with_dir(here::here("docs"), knitr::include_graphics("./figures/example.png"))`

Ideally, however, I would like the `docs/figures` directory to be a "target", so that I can see the dependency between my report and the figure directory in my pipeline. The way I was doing this was by providing the directory as an external parameter to my `.rmd` (see [here](https://github.com/RHReynolds/rtargets-reprex/blob/f949fb0c89744dd896bf7606a14e3a30cf7260cf/docs/example.rmd#L5)), having this directory be a target within `_targets.R`, and then calling the figure with `knitr::include_graphics(file.path(params$ext_figures, "example.png"))` in the rmd.  This may be very obvious, and I'm just not seeing it, but how would I get the `withr` solution to work such that I could monitor these dependencies?

I've created a GitHub repo that should help with the reprex: [rtargets-reprex](https://github.com/RHReynolds/rtargets-reprex).

**Reply by wlandau** [reply]
Thanks for the reprex. I put together a recommendation at https://github.com/RHReynolds/rtargets-reprex/pull/1. The main changes I proposed are:

* Set the `knitr` root directory to be the pipeline root.
* Set an option to make `knitr::include_graphics()` respect absolute paths.
* Supply the absolute path to the figure to  `knitr::include_graphics()`.

I also modified the solution to work without `here::here()` (for users who may not be using `here`) and simplified it to work without R Markdown parameters. The `docs/example.rmd` report now looks like this:

    ---
    title: "Example"
    date: "`r format(Sys.time(), '%d %B %Y')`"
    output: 
      bookdown::html_document2:
        figure_caption: yes
        code_folding: hide
        theme: paper
        highlight: kate
        df_print: paged
        toc: true
        toc_float: true
        number_sections: true
    ---
    
    ```{r setup}
    root <- rprojroot::find_root_file(criterion = "_targets.R") # Find the pipeline root directory.
    knitr::opts_knit$set(root.dir = root) # Run the report from the pipeline root directory. 
    options(knitr.graphics.rel_path = FALSE) # Make knitr::include_graphics() use absolute paths.
    ```
    
    ```{r example-figure, fig.cap = "Example figure"}
    knitr::include_graphics(file.path(root, targets::tar_read(fig_dir), "example.png"))
    ```

**Reply by RHReynolds** [reply]
Thank you very much for this!!
